Sprint 05

Rozetka220 opened the pull request:
Это было очень и очень сложно.  Особенно делегаты.

Rozetka220 committed:
sprint_4

Rozetka220 committed:
delete comments

Rozetka220 committed:
Change Lounge Screen content mode to aspect fill
Change constraint general stack in main file = safe area bottom(0)
Delete padding question label constraint
Change auto shrink fixed -> minimum font size = 18

Rozetka220 committed:
fix extra correct answers
add disable "yes" and "no" buttons

Rozetka220 committed:
Вроде блеять все делаю как написано в этой ебучей теории, но нихуя сука не работает, помогите((((

Rozetka220 committed:
before sprint_05 task

Rozetka220 committed:
Спринт 5. Задание по теме.
Я абсолютно не понимаю как, но кажется это работает...

Rozetka220 committed:
The end 5-spint

helecta commented on MovieQuiz/Models/QuizResultsViewModel.swift:
```diff
@@ -0,0 +1,15 @@
+
+import Foundation
+
+// для состояния "Результат квиза"
```
Здорово, что добавляешь комментарии. Если добавлять через /// , а не // , то тогда можно будет быстро прочитать документацию, если зажать кнопку option и навести на класс.

Rozetka220 replied:
Даже и подумать не мог, хотя замечал, что комментарий с "///" тоже существует. Спасибо!

helecta commented on MovieQuiz/Presentation/AlertPresenter.swift:
```diff
@@ -0,0 +1,34 @@
+import UIKit
+
+class AlertPresenter: AlertPresenterProtocol {
+    internal weak var delegateAlert = UIViewController()
```
Для того чтобы показать алерт, не нужно специально создавать новый ViewController. Достаточно передавать контроллер, на котором необходимо показать алерт, сюда как делегат. Закомментированный init это верный ход рассуждений

helecta commented on MovieQuiz/Services/AlertPresenterProtocol.swift:
```diff
@@ -0,0 +1,15 @@
+
+protocol AlertPresenterProtocol {
+    //func convertToAlertModel(model: QuizResultsViewModel) -> AlertModel
+    var delegateAlert: UIViewController? {get set}
```
Название delegateAlert не удачное, обычно слово delegate должно быть в конце.

helecta commented on MovieQuiz/Services/QuestionFactoryProtocol.swift:
```diff
@@ -0,0 +1,12 @@
+
+import Foundation
+
+protocol QuestionFactoryProtocol : QuestionFactoryDelegate {
```
Почему он QuestionFactoryDelegate?

Rozetka220 replied:
<img width="1294" alt="изображение" src="https://user-images.githubusercontent.com/16082900/211927714-bf83eb0f-6a48-4571-b480-2f72df86ab29.png">
Видимо для этого, я уже не помню, зачем так было нужно. На сколько критично, если я это оставлю как есть?

helecta replied:
Это довольно важный момент. Вопрос был скорее наводящим) 
Попробуй разобраться, какая ответственность должна быть у QuestionFactoryProtocol и какая у QuestionFactoryDelegate.
Если возникнут сложности, пиши либо тут либо Наставнику.
P.S следующий комментарий к QuestionFactoryDelegate.swift в общем-то должен немного помочь

Rozetka220 replied:
Хорошо, попробую сегодня разобраться. Моя вчерашняя попытка завершилась крахом с кучей ошибок.  Перечитав с утра так и подумал, что ключом к разгадке может быть следующий комментарий. Спасибо!

Rozetka220 replied:
Я думаю, ответственность такая: 
QuestionFactoryProtocol - нужен, что бы получать вопросы из массива вопросов (хранилища вопросов)

QuestionFactoryDelegate - нужен, что бы этот вопрос можно было передать в viewController

viewController - отображает вопрос на экране

Rozetka220 replied:
Никак не могу понять, как про инициализировать _weak var delegate: QuestionFactoryDelegate?_

В учебнике написано, что инициализация происходит через свойство, однако это больше похоже на первый вариант из тема "**Делегаты**" (второй скриншот)
<img width="963" alt="изображение" src="https://user-images.githubusercontent.com/16082900/212118630-c3001733-9dd2-4324-9067-c007d321764a.png">

<img width="1440" alt="изображение" src="https://user-images.githubusercontent.com/16082900/212119193-4fa1f6d0-1a99-4afd-be2f-dbb2da7ff93b.png">


В итоге у меня выходит ошибка "_Argument passed to call that takes no arguments_", что кажется логичным, ведь конструктора на свойство _delegate_ у меня нет

<img width="989" alt="изображение" src="https://user-images.githubusercontent.com/16082900/212120942-f9332192-5f22-4394-91b9-318268b8bbd3.png">

 
Если я попробую про инициализировать через свойство так, как я понима, то у меня выходит ошибка "_Value of type 'any QuestionFactoryProtocol' has no member 'delegate'_" И ведь правда, переменная делегата в _MovieQuizViewController_ типа _QuestionProtocol_, а в нем у меня нет такого свойства

<img width="961" alt="изображение" src="https://user-images.githubusercontent.com/16082900/212120615-9f3ffaba-d173-4894-9792-6476ede82818.png">

Rozetka220 replied:
В итоге выбрал инициализацию через инициализацию и добавил конструктор в _QuestionFactory_, но я не помню, что бы в учебнике это было и не отхожу ли я так от рамок, в которых я должен справиться с задачей?
Говорю, что не помню, потому что уже нет сил перечитывать этот спринт и меня от него тошнит, и проверять, действительно ли там не было конструктора или просто я где-то проглядел и накосячил. Это не правильно, но я ничего не могу с собой поделать
<img width="650" alt="изображение" src="https://user-images.githubusercontent.com/16082900/212121892-233f5e2a-3fd3-4954-9495-0e70814ce2a4.png">

Rozetka220 replied:
В общем, буду ждать тут от тебя ответа, все остальное исправил, по крайней мере мне так кажется. Если все верно, сразу скину проект на перепроверку. Надеюсь, завтра вечером или хотя бы ночью, я уже смогу приступить к новому спринту.

Rozetka220 replied:
Отправил, если надо будет что-то исправить, исправлю и перезалью

helecta replied:
Да, действительно в учебнике не совсем ясно сказано. 
Как я поняла, здесь:
> Мы реализовали инъекцию через свойство. Поэтому задавать делегата мы будет во viewDidLoad методе.

речь была об инъекции через свойство questionFactory для MovieQuizViewController.
Т.е инъекции через свойство происходит во viewDidLoad:
`questionFactory = QuestionFactory(delegate: self)`,
а если бы была инъекции через инициализацию, то она выглядела бы так:
`private var questionFactory: QuestionFactoryProtocol? =  QuestionFactory(delegate: self)`

Сейчас ты сделал все как надо

helecta commented on MovieQuiz/Services/QuestionFactoryDelegate.swift:
```diff
@@ -0,0 +1,13 @@
+import Foundation
+
+protocol QuestionFactoryDelegate : AnyObject {
+    var delegate: QuestionFactoryDelegate? { get set }
```
Тут не нужен этот delegate. 
У самой фабрики есть делегат ` weak var delegate: QuestionFactoryDelegate?`, который занимается обработкой момента получения следующего вопроса( в нашем случае делегатом будет viewController, который и займется обработкой этого момента)

helecta commented on MovieQuiz/Services/StatisticService.swift:
```diff
@@ -0,0 +1,105 @@
+        case correct, total, bestGame, gamesCount
+    }
+    
+    var getSetTotalAccuracy: Double {
```
Свойства не принято так называть, ошибка что тут есть глаголы getSet. 
Аналогично нужно исправить и все остальные свойства

Rozetka220 replied:
Могу ли я сделать просто вот так? 
<img width="617" alt="изображение" src="https://user-images.githubusercontent.com/16082900/211926502-c466cea3-fd7f-4387-8cb8-ca5d0f0e3760.png">

helecta replied:
Да, так и нужно

helecta commented on MovieQuiz/Services/StatisticService.swift:
```diff
@@ -0,0 +1,105 @@
+    var getSetTotalAccuracy: Double {
+        get{
+            guard let data = userDefaults.data(forKey: Keys.total.rawValue),
+                  let totalAccuracy = try? JSONDecoder().decode(Double.self, from: data) else {
```
Тут можно упростить, обрати внимание UserDefaults поддерживает работу с примитивными типами [для чтения](https://developer.apple.com/documentation/foundation/userdefaults/1407405-integer) и [записи](https://developer.apple.com/documentation/foundation/userdefaults/1413614-set).

helecta commented on MovieQuiz/Services/StatisticService.swift:
```diff
@@ -0,0 +1,105 @@
+    
+    func store(correct count: Int, total amount: Int) {
+        //здесь должны записывать новый лучший результат в бестгейм
+        //refresh()
```
Перед сдачей стоит убирать весь закомментированный код.

helecta commented on MovieQuiz/Services/StatisticService.swift:
```diff
@@ -0,0 +1,105 @@
+        print("count = \(count) total = \(amount) totalAccuracy = \(totalAccuracy)")
+    }
+    
+    func refresh(){
```
Этот метод нигде не используется(кроме как в закомментированном коде), поэтому его стоит удалить

helecta commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -1,72 +1,147 @@
+    private func showNextQuestionOrResults() {
+        if currentQuestionIndex == questionsAmount - 1 {
+            serviceStatictic.store(correct: correctAnswers, total: questionsAmount)
+            let text = "Ваш результат: \(correctAnswers) из \(questionsAmount)\nКоличество сыгранных квизов: \(serviceStatictic.gamesCount)\nРекорд: \(serviceStatictic.bestGame.correct)/\(questionsAmount) \(serviceStatictic.bestGame.date.dateTimeString)\nСредняя точность: \(String(format: "%.2f", serviceStatictic.totalAccuracy))%"
```
Можно использовать [многострочный вариант строки](https://www.swiftbysundell.com/tips/multiline-string-literal-tips-and-tricks/), чтобы строчка кода умещалась в столбец

Rozetka220 replied:
Не совсем понял смысл. Он в том, что в коде все выглядело более опрятно и читаемо?
<img width="934" alt="изображение" src="https://user-images.githubusercontent.com/16082900/211923502-4d94afb0-7232-4f0f-acc6-1f2937944e06.png">

helecta replied:
Да, все верно. Так просто лучше читаемость кода

helecta commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -1,72 +1,147 @@
+    private func showNextQuestionOrResults() {
+        if currentQuestionIndex == questionsAmount - 1 {
+            serviceStatictic.store(correct: correctAnswers, total: questionsAmount)
+            let text = "Ваш результат: \(correctAnswers) из \(questionsAmount)\nКоличество сыгранных квизов: \(serviceStatictic.gamesCount)\nРекорд: \(serviceStatictic.bestGame.correct)/\(questionsAmount) \(serviceStatictic.bestGame.date.dateTimeString)\nСредняя точность: \(String(format: "%.2f", serviceStatictic.totalAccuracy))%"
```
Обрати внимание на макет. Там в тексте "Ваш результат" дальше нет "из", используется формат "n/10"

helecta commented on MovieQuiz/Services/AlertPresenterProtocol.swift:
```diff
@@ -0,0 +1,15 @@
+protocol AlertPresenterProtocol {
+    //func convertToAlertModel(model: QuizResultsViewModel) -> AlertModel
+    var delegateAlert: UIViewController? {get set}
+    func requestGameOverBtn(model: AlertModel)
```
Не удачное название метода. 
Для того чтобы понять, как лучше называть методы, можно задать вопрос, что делает AlertPresenter, за что отвечает этот класс?
В нашем случае его ответственность - это показывает алерт, поэтому и метод стоит назвать так.

helecta commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -1,72 +1,147 @@
+    
+    //var delegateAlert: AlertPresenterProtocol = AlertPresenter()
+    
+    func requestGameOverBtn(model: AlertModel){}
```
Методы стоит перенести к методам, даже если они пустые

Rozetka220 replied:
И вправду, это результат борьбы, когда xcode принимал соответствие протоколу только через Fix, я потом побоялся что-то трогать

helecta commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -1,72 +1,147 @@
+        }
+    }
+        
+    @IBAction func yesButton(_ sender: Any) {
```
Это метод, поэтому в название стоит использовать глагол, например Pressed

helecta commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -1,72 +1,147 @@
+        currentQuestion.correctAnswer ? showAnswerResult(isCorrect: true) : showAnswerResult(isCorrect: false)
+    }
+    
+    @IBAction func noButton(_ sender: Any) {
```
Аналогично

Rozetka220 committed:
Fix many comments from my reviewer:
-rename functions and variables
-remove delegate костыли (QuestionFactory, QuestionFactoryDelegate protocols)
	-add init in class QuestionFactory

helecta commented on MovieQuiz/Presentation/AlertPresenter.swift:
```diff
@@ -9,10 +9,10 @@ import Foundation
     
-    func requestGameOverBtn(model: AlertModel) {
-        guard let delegateAlert = delegateAlert else {return}
+    func viewGameOverAlert(model: AlertModel) {
```
Лучше назвать не `viewGameOverAlert`, а `presentGameOverAlert`.
View в название может ввести в заблуждение, что это UIView.
Present - обычно означает показать поверх текущего View
Show - вместо текущего View

Rozetka220 replied:
Понял, принял, спасибо!

helecta commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -1,11 +1,7 @@
 import UIKit
 
 final class MovieQuizViewController: UIViewController, QuestionFactoryDelegate, AlertPresenterProtocol {
```
Возможно забыл убрать, зачем MovieQuizViewController отвечает протоколу AlertPresenterProtocol?
Если убрать, то не забудь убрать и
`var alertDelegate: UIViewController?`
`func viewGameOverAlert(model: AlertModel){}`

Rozetka220 replied:
Возможно. Очень сильно мучился с эти заданием.
